snuba/datasets/schemas/tables.py

```diff
@@ -255,6 +261,33 @@ def __init__(
     def _get_engine_type(self) -> str:
         return "ReplacingMergeTree(%s)" % self.__version_column
 
+    def get_required_columns(self) -> Sequence[str]:
+        """
+        Returns the list of the columns of this table schema that cannot be left
+        empty when writing an INSERT query.
+
+        This delas with flattened columns (since the default iteration behavior on
+        ColumnSet is to return flattened columns, thus this approach seems more robust).
+        This means there are no Nested columns, only Arrays.
+        """
```

Comment:
I don't think this is necessarily accurate, [looking at the documentation](https://clickhouse.tech/docs/en/query_language/insert_into/):

> The query can specify a list of columns to insert [(c1, c2, c3)]. In this case, the rest of the columns are filled with:
> - The values calculated from the DEFAULT expressions specified in the table definition.
> - Zeros and empty strings, if DEFAULT expressions are not defined.

This doesn't say anything about arrays (looking at the comment below) but I assume if there is no default expression, they just become empty. It doesn't actually say anything about `Nullable` columns either, but I assume those would become `NULL` instead of the non-`NULL` default.

With this in mind (assuming we're not running with `strict_insert_defaults` enabled), I don't think whether or not a column is `Nullable` has a significant impact on whether or not it should be included in columns that are preserved in the deletion query. In fact, if we left out `group_id` from the columns to preserve during deletion, it would have actually ended up doing the thing that I mentioned [in this comment](https://github.com/getsentry/snuba/pull/778#discussion_r379600079), which makes me even more confused about how it ended up in there in the first place.[1]

I think that the only columns that need to be preserved here during are the columns that affect the viability of a row to be merged during optimization — specifically rows that are referenced in the `ORDER BY` and `PARTITION BY` DDL clauses.

In retrospect, I realize the "this table contains far less nullable fields" part of [this comment](https://github.com/getsentry/snuba/pull/778#discussion_r379605521) was a bit misleading because I misunderstood how non-`Nullable` columns are treated with `INSERT INTO` and expected it to behave more like PostgreSQL or MySQL.

1: One weird side effect of these replacements is that they will trash the min side of a min/max index since they are not conditional on `deleted = 1`.

Reply:
damn, that's why we did not receive complaints of replacements not working after introducing flattened_tags. The column is not nullable, we were not providing anything so it was adding the default .....

```
I think that the only columns that need to be preserved here during are the columns that affect the viability of a row to be merged during optimization — specifically columns that are referenced in the ORDER BY and PARTITION BY DDL clauses.
```

Those plus the deleted column. So I guess I will go back to the original implementation with the only difference of splitting order by columns and partition by columns.